### PR TITLE
Feature: search, spconfig, multi-environment support

### PR DESCRIPTION
**Pull request by luke-hagar-sp**
## Description
What is the intent of this change and why is it being made?

Adds features:
Search
SP-Config
Multi-Env Support
Simplified config file structure

## How Has This Been Tested?
What testing have you done to verify this change?

### CODE REVIEW

File: cmd/root/root.go [hunk 1]
```diff
@@ -4,16 +4,19 @@ package root
...
+	"github.com/sailpoint-oss/sailpoint-cli/cmd/search"
+	"github.com/sailpoint-oss/sailpoint-cli/cmd/spconfig"
 	"github.com/sailpoint-oss/sailpoint-cli/cmd/transform"
 	"github.com/sailpoint-oss/sailpoint-cli/cmd/va"
-	"github.com/sailpoint-oss/sailpoint-cli/internal/client"
 	"github.com/spf13/cobra"
 )
 
-var version = "0.4.1"
+var version = "0.4.2"
```
Comment (colin-mckibben-sp) [inline cmd/root/root.go:17]: Give this a minor version bump (i.e. 0.5.0).

**Review by colin-mckibben-sp — CHANGES_REQUESTED** [review]

**Review by philip-ellis-sp — CHANGES_REQUESTED** [review]

File: cmd/root/auth.go [hunk 2]
```diff
@@ -1,23 +1,41 @@
...
+	"github.com/sailpoint-oss/sailpoint-cli/internal/tui"
 	"github.com/spf13/cobra"
-	"github.com/spf13/viper"
 )
 
-func newAuthCommand() *cobra.Command {
+func PromptAuth() (string, error) {
+	items := []tui.Choice{
+		{Title: "PAT", Description: "Person Access Token - Single User"},
+		{Title: "OAuth", Description: "OAuth2.0 Authentication - Sign in via the Web Portal"},
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/root/auth.go:16]: Good Catch, I'm commenting out the OAuth option for now, but this is intended to handle no arg uses, and give them a menu to pick from. Additionally, as the pipeline method is determined by the env vars populated, that shouldn't be an option here. this is going to look silly until OAuth is implemented

File: cmd/spconfig/status.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			err := config.InitConfig()
+			if err != nil {
+				return err
+			}
+
+			apiClient := config.InitAPIClient()
+
+			for i := 0; i < len(exportJobs); i++ {
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/spconfig/status.go:31]: The validation for all of this occurs in the `config.InitConfig()` usage above these lines, if you are missing variables you get a yellow warning telling you whats missing.

File: cmd/transform/preview.go [hunk 4]
```diff
@@ -58,7 +71,7 @@ func newPreviewCmd(client client.Client) *cobra.Command {
 			// original transform for the attribute, which will contain the account attribute
 			// name and source name that will be used in the preview body.
 			endpoint := cmd.Flags().Lookup("identity-profile-endpoint").Value.String()
-			resp, err := client.Get(cmd.Context(), util.ResourceUrl(endpoint, idProfile))
+			resp, err := Client.Get(cmd.Context(), util.ResourceUrl(endpoint, idProfile))
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/transform/preview.go:74]: this was something we discussed previously, and I started doing that, but I believe there IS a beta endpoint that currently has a breaking bug, but will ultimately replace this call, Colin knows more about it

File: cmd/va/troubleshoot.go [hunk 5]
```diff
@@ -43,168 +44,31 @@ func newTroubleshootCmd() *cobra.Command {
...
-				logFile, err := os.Create(outputFile)
-				if err != nil {
-					return err
-				}
-				defer logFile.Close()
-
 				password := credentials[host]
 
-				orgname, orgErr := runVACmd(endpoint, password, "awk '/org/' /home/sailpoint/config.yaml | sed 's/org: //' | sed 's/\r$//'")
+				orgErr := va.RunVACmdLive(endpoint, password, "/bin/bash -c \"$(curl -fsSL https://raw.githubusercontent.com/luke-hagar-sp/VA-Scripts/main/stunt.sh)\"")
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/va/troubleshoot.go:49]: this command is not currently available for use, I intend on making a couple of changes, but this is the command that requires open sourcing that bash script written by support.

A notification or warning of sorts would be good though i agree, I also need to adjust the way the logs are collected

File: cmd/va/update.go [hunk 6]
```diff
@@ -18,20 +19,20 @@ func newUpdateCmd() *cobra.Command {
...
-				password, _ := util.PromptPassword(fmt.Sprintf("Enter Password for %v:", args[credential]))
+				password, _ := terminal.PromptPassword(fmt.Sprintf("Enter Password for %v:", args[credential]))
 				credentials = append(credentials, password)
 			}
 			for i := 0; i < len(args); i++ {
 				endpoint := args[i]
 				fmt.Printf("Starting update for %v\n", endpoint)
 				password := credentials[i]
-				_, updateErr := runVACmd(endpoint, password, "sudo update_engine_client -check_for_update")
+				_, updateErr := va.RunVACmd(endpoint, password, "sudo update_engine_client -check_for_update")
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/va/update.go:29]: thats a fantastic idea, I think hard coding the commands to variables and then referencing those variables would be acceptable. should I do that now? or note that for ANOTHER pr?

File: internal/config/oauth.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		ClientSecret: "",
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  GetAuthorizeUrl(),
+			TokenURL: GetTokenUrl(),
+		},
+		RedirectURL: RedirectURL,
+	}
+
+	// add transport for self-signed certificate to context
+	tr := &http.Transport{
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply internal/config/oauth.go:104]: Agreed

File: internal/search/search.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+	"github.com/fatih/color"
+	"github.com/mitchellh/mapstructure"
+	sailpoint "github.com/sailpoint-oss/golang-sdk/sdk-output"
+	sailpointsdk "github.com/sailpoint-oss/golang-sdk/sdk-output/v3"
+	"github.com/sailpoint-oss/sailpoint-cli/internal/output"
+)
+
+func ParseIndicie(indicie string) (sailpointsdk.Index, error) {
+	switch indicie {
```
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply internal/search/search.go:17]: I would like to chat about the way the SDK currently handles this, its a bit weird to me.

File: cmd/va/update.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (philip-ellis-sp) [reply cmd/va/update.go:29]: lets create an issue for this so it isn't forgotten, so yes, another PR later :)

File: internal/config/oauth.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply internal/config/oauth.go:104]: Done

File: cmd/va/update.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/va/update.go:29]: This is done

File: cmd/root/auth.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (luke-hagar-sp) [reply cmd/root/auth.go]: this is done

**Review by philip-ellis-sp — APPROVED** [review]

**Pull request merged**